linkerd/app/outbound/src/http/logical/policy.rs

```diff
@@ -0,0 +1,396 @@
+use super::{super::concrete, BackendCache, Concrete, Distribution, NoRoute};
+use linkerd_app_core::{
+    proxy::http::{self, balance},
+    svc,
+    transport::addrs::*,
+    Addr, Error, Infallible,
+};
+use linkerd_distribute as distribute;
+use linkerd_http_route as route;
+use linkerd_proxy_client_policy as policy;
+use std::{fmt::Debug, hash::Hash, sync::Arc};
+
+#[cfg(test)]
+mod tests;
+
+pub use linkerd_proxy_client_policy::ClientPolicy;
+
+/// HTTP or gRPC policy routes.
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub enum Routes {
+    Http(HttpRoutes),
+    Grpc(GrpcRoutes),
+}
+
+pub type HttpRoutes = RouterRoutes<route::http::MatchRequest, policy::http::Filter>;
+pub type GrpcRoutes = RouterRoutes<route::grpc::MatchRoute, policy::grpc::Filter>;
+
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub struct RouterRoutes<M, F> {
+    pub addr: Addr,
+    pub routes: Arc<[route::Route<M, policy::RoutePolicy<F>>]>,
+    pub backends: Arc<[policy::Backend]>,
+}
+
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub(super) enum Params<T: Clone + Debug + Eq + Hash> {
+    Http(HttpParams<T>),
+    Grpc(GrpcParams<T>),
+}
+
+pub(super) type HttpParams<T> = RouterParams<T, route::http::MatchRequest, policy::http::Filter>;
+pub(super) type GrpcParams<T> = RouterParams<T, route::grpc::MatchRoute, policy::grpc::Filter>;
+
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub(super) struct RouterParams<T: Clone + Debug + Eq + Hash, M, F> {
+    parent: T,
+    addr: Addr,
+    routes: Arc<[route::Route<M, RouteParams<T, F>>]>,
+    backends: distribute::Backends<Concrete<T>>,
+}
+
+#[derive(Clone, Debug, PartialEq, Eq, Hash)]
+pub(super) struct RouteParams<T, F> {
+    parent: T,
+    addr: Addr,
+    meta: Arc<policy::Meta>,
+    filters: Arc<[F]>,
+    distribution: Distribution<T>,
+}
+
+// === impl Routes ===
+
+impl Routes {
+    pub fn addr(&self) -> &Addr {
+        match self {
+            Routes::Http(RouterRoutes { ref addr, .. })
+            | Routes::Grpc(RouterRoutes { ref addr, .. }) => addr,
+        }
+    }
+}
+
+// === impl Params ===
+
+impl<T> Params<T>
+where
+    // Parent target type.
+    T: Clone + Debug + Eq + Hash + Send + Sync + 'static,
+{
+    /// HTTP or gRPC policy routing.
```

Comment:
nit, TIOLI: maybe
```suggestion
    /// Wraps an HTTP `NewService` with HTTP or gRPC policy routing layers.
```
or something?